Saga metadata a first level concept

johnsimons opened the pull request:
## Changes included in this PR

1. `UniqueAttribute` is now deprecated, correlation properties are automatically *unique*, what this means is that NServiceBus will automatically uniquely correlate the properties mapped, eg:

    ```c#
    protected override void ConfigureHowToFindSaga(SagaPropertyMapper<Saga1Data> mapper)
    {
        mapper.ConfigureMapping<StartSaga>(m => m.DataId).ToSaga(s => s.DataId);
    }
    ```
In the above example `DataId` will be automatically correlated.
1. `NServiceBus.Sagas.ActiveSagaInstance` moved to `NServiceBus.Saga.ActiveSagaInstance`
1. `ISagaPersister` interface was changed to include `void Initialize(SagaMetaModel model)`, this new method is invoked by the framework to provide implementation classes with the new `SagaMetaModel` model, which exposes all the metamodel for a saga type.
1. A new `SagaMetadata` class is now exposed that provide advance users with the ability to query the metadata model for saga details, such as:
    - `CorrelationProperties`, list of mappings;
    - Finders, this is the list of `IFindSagas<T>.Using<M>`;
    - `SagaEntityType`;
      
   The `SagaMetadata` instance is exposed via `ISagaPersister.Initialize`

## Issues addressed

andreasohlund commented on src/NServiceBus.Core/Sagas/CustomFinderAdapter.cs:
```diff
@@ -1,8 +1,7 @@
-namespace NServiceBus.Sagas
+namespace NServiceBus.Saga
```
Won't `Saga` clash with our `Saga<T>` type?

SimonCropp replied:
is there really any value in obscuring this api under a nested namespace?

andreasohlund replied:
Since it's not relevant to end users don't we want to keep it out of the
root namespace?

On Thu, Mar 26, 2015 at 11:20 AM, Simon Cropp <notifications@github.com>
wrote:

> In src/NServiceBus.Core/Sagas/CustomFinderAdapter.cs
> <https://github.com/Particular/NServiceBus/pull/2693#discussion_r27201036>
> :
>
> > @@ -1,8 +1,7 @@
> > -namespace NServiceBus.Sagas
> > +namespace NServiceBus.Saga
>
> is there really any value in obscuring this api under a nested namespace?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/2693/files#r27201036>.
>

SimonCropp replied:
ok. just checking. but yes namespaces with the same name as types is a pain

johnsimons replied:
@andreasohlund both `Saga` and `Saga<T>` are already in `NServiceBus.Saga` namespace, so I did not change that.
All I am doing here is to standardising on a single namespace for the the saga stuff.

Makes sense ?

andreasohlund replied:
Makes alot of sense. But I think `Sagas` is a better namespace name?

johnsimons replied:
I agree but all public classes are currently siting in Saga so moving all
those to sagas is a much bigger change

On Friday, March 27, 2015, Andreas Öhlund <notifications@github.com> wrote:

> In src/NServiceBus.Core/Sagas/CustomFinderAdapter.cs
> <https://github.com/Particular/NServiceBus/pull/2693#discussion_r27279193>
> :
>
> > @@ -1,8 +1,7 @@
> > -namespace NServiceBus.Sagas
> > +namespace NServiceBus.Saga
>
> Makes alot of sense. But I think Sagas is a better namespace name?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/2693/files#r27279193>.
>

johnsimons replied:
Not worth the effort IMO

On Friday, March 27, 2015, John Simons <john.simons@particular.net> wrote:

> I agree but all public classes are currently siting in Saga so moving all
> those to sagas is a much bigger change
>
> On Friday, March 27, 2015, Andreas Öhlund <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> In src/NServiceBus.Core/Sagas/CustomFinderAdapter.cs
>> <https://github.com/Particular/NServiceBus/pull/2693#discussion_r27279193>
>> :
>>
>> > @@ -1,8 +1,7 @@
>> > -namespace NServiceBus.Sagas
>> > +namespace NServiceBus.Saga
>>
>> Makes alot of sense. But I think Sagas is a better namespace name?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/Particular/NServiceBus/pull/2693/files#r27279193>.
>>
>

andreasohlund replied:
I propose we move the saga base class to the root `NServiceBus` namespace
to be more discoverable OOTB. If we do that why should we keep the
`NServiceBus.Saga` namespace?

If we do we'll have both a class and a namespace with the same name == pain?

On Fri, Mar 27, 2015 at 8:28 AM, John Simons <notifications@github.com>
wrote:

> In src/NServiceBus.Core/Sagas/CustomFinderAdapter.cs
> <https://github.com/Particular/NServiceBus/pull/2693#discussion_r27279285>
> :
>
> > @@ -1,8 +1,7 @@
> > -namespace NServiceBus.Sagas
> > +namespace NServiceBus.Saga
>
> Not worth the effort IMO
> … <#14c5a219cec958a8_>
> On Friday, March 27, 2015, John Simons <john.simons@particular.net>
> wrote: > I agree but all public classes are currently siting in Saga so
> moving all > those to sagas is a much bigger change > > On Friday, March
> 27, 2015, Andreas Öhlund <notifications@github.com >
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote: > In
> src/NServiceBus.Core/Sagas/CustomFinderAdapter.cs > <
> https://github.com/Particular/NServiceBus/pull/2693#discussion_r27279193>
> > : > > > @@ -1,8 +1,7 @@ > > -namespace NServiceBus.Sagas > > +namespace
> NServiceBus.Saga > > Makes alot of sense. But I think Sagas is a better
> namespace name? > > — > Reply to this email directly or view it on GitHub >
> <https://github.com/Particular/NServiceBus/pull/2693/files#r27279193>. >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/2693/files#r27279285>.
>

johnsimons replied:
> If we do we'll have both a class and a namespace with the same name ==
pain?

There is no pain, as I said before that is how it is currently out in prod.

On Friday, March 27, 2015, Andreas Öhlund <notifications@github.com> wrote:

> In src/NServiceBus.Core/Sagas/CustomFinderAdapter.cs
> <https://github.com/Particular/NServiceBus/pull/2693#discussion_r27279405>
> :
>
> > @@ -1,8 +1,7 @@
> > -namespace NServiceBus.Sagas
> > +namespace NServiceBus.Saga
>
> I propose we move the saga base class to the root `NServiceBus` namespace
> to be more discoverable OOTB. If we do that why should we keep the
> `NServiceBus.Saga` namespace? If we do we'll have both a class and a
> namespace with the same name == pain?
> … <#14c5a26a2031e22c_>
> On Fri, Mar 27, 2015 at 8:28 AM, John Simons <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote: In
> src/NServiceBus.Core/Sagas/CustomFinderAdapter.cs <
> https://github.com/Particular/NServiceBus/pull/2693#discussion_r27279285>
> : > @@ -1,8 +1,7 @@ > -namespace NServiceBus.Sagas > +namespace
> NServiceBus.Saga Not worth the effort IMO … <#14c5a219cec958a8_> On Friday,
> March 27, 2015, John Simons <john.simons@particular.net
> <javascript:_e(%7B%7D,'cvml','john.simons@particular.net');>> wrote: > I
> agree but all public classes are currently siting in Saga so moving all >
> those to sagas is a much bigger change > > On Friday, March 27, 2015,
> Andreas Öhlund <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');> >
> <javascript:_e(%7B%7D,'cvml','notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>');>> wrote: >
> In src/NServiceBus.Core/Sagas/CustomFinderAdapter.cs > <
> https://github.com/Particular/NServiceBus/pull/2693#discussion_r27279193>
> > : > > > @@ -1,8 +1,7 @@ > > -namespace NServiceBus.Sagas > > +namespace
> NServiceBus.Saga > > Makes alot of sense. But I think Sagas is a better
> namespace name? > > — > Reply to this email directly or view it on GitHub >
> <https://github.com/Particular/NServiceBus/pull/2693/files#r27279193>. >
> — Reply to this email directly or view it on GitHub <
> https://github.com/Particular/NServiceBus/pull/2693/files#r27279285>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/2693/files#r27279405>.
>

andreasohlund replied:
Are you saying that it's not valid to move the `Saga`base class to
`NServiceBus` to make it more discoverable?

On Fri, Mar 27, 2015 at 8:35 AM, John Simons <notifications@github.com>
wrote:

> In src/NServiceBus.Core/Sagas/CustomFinderAdapter.cs
> <https://github.com/Particular/NServiceBus/pull/2693#discussion_r27279449>
> :
>
> > @@ -1,8 +1,7 @@
> > -namespace NServiceBus.Sagas
> > +namespace NServiceBus.Saga
>
>  If we do we'll have both a class and a namespace with the same name ==
> pain? There is no pain, as I said before that is how it is currently out
> in prod.
> … <#14c5a2849eaabf24_>
> On Friday, March 27, 2015, Andreas Öhlund <notifications@github.com>
> wrote: In src/NServiceBus.Core/Sagas/CustomFinderAdapter.cs <
> https://github.com/Particular/NServiceBus/pull/2693#discussion_r27279405>
> : > @@ -1,8 +1,7 @@ > -namespace NServiceBus.Sagas > +namespace
> NServiceBus.Saga I propose we move the saga base class to the root
> `NServiceBus` namespace to be more discoverable OOTB. If we do that why
> should we keep the `NServiceBus.Saga` namespace? If we do we'll have both a
> class and a namespace with the same name == pain? … <#14c5a26a2031e22c_> On
> Fri, Mar 27, 2015 at 8:28 AM, John Simons <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote: In
> src/NServiceBus.Core/Sagas/CustomFinderAdapter.cs <
> https://github.com/Particular/NServiceBus/pull/2693#discussion_r27279285>
> : > @@ -1,8 +1,7 @@ > -namespace NServiceBus.Sagas > +namespace
> NServiceBus.Saga Not worth the effort IMO … <#14c5a219cec958a8_> On Friday,
> March 27, 2015, John Simons <john.simons@particular.net
> <javascript:_e(%7B%7D,'cvml','john.simons@particular.net');>> wrote: > I
> agree but all public classes are currently siting in Saga so moving all >
> those to sagas is a much bigger change > > On Friday, March 27, 2015,
> Andreas Öhlund <notifications@github.com <javascript:_e(%7B%7D,'cvml','
> notifications@github.com');> > <javascript:_e(%7B%7D,'cvml','
> notifications@github.com <javascript:_e(%7B%7D,'cvml','
> notifications@github.com');>');>> wrote: > In
> src/NServiceBus.Core/Sagas/CustomFinderAdapter.cs > <
> https://github.com/Particular/NServiceBus/pull/2693#discussion_r27279193>
> > : > > > @@ -1,8 +1,7 @@ > > -namespace NServiceBus.Sagas > > +namespace
> NServiceBus.Saga > > Makes alot of sense. But I think Sagas is a better
> namespace name? > > — > Reply to this email directly or view it on GitHub >
> <https://github.com/Particular/NServiceBus/pull/2693/files#r27279193>. >
> — Reply to this email directly or view it on GitHub <
> https://github.com/Particular/NServiceBus/pull/2693/files#r27279285>. —
> Reply to this email directly or view it on GitHub <
> https://github.com/Particular/NServiceBus/pull/2693/files#r27279405>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/2693/files#r27279449>.
>

johnsimons replied:
> Are you saying that it's not valid to move the `Saga`base class to
> `NServiceBus` to make it more discoverable?

Yes, I don't think is worth the hassle, but up to you

andreasohlund replied:
Ok, I'll do this as a separate pull after this one is in

johnsimons commented:
@andreasohlund @danielmarbach as we spoke about it, I have added an `Initialize(SageMetaModel)` to `ISagaPersister', I've modified the NH impl, but NH does not require it at all, so this method impl for NH is a noop.

Looking at the code for Raven impl (https://github.com/Particular/NServiceBus.RavenDB/blob/develop/src/NServiceBus.RavenDB/SagaPersister/SagaPersister.cs) it looks like the `Save`, `Update`, `Get`, `Complete` methods need to do a lookup by sagadata type, not Saga type:
- https://github.com/Particular/NServiceBus.RavenDB/blob/70fd44638395148d9acbcb01a1e2138f930a555c/src/NServiceBus.RavenDB/SagaPersister/SagaPersister.cs#L22
- https://github.com/Particular/NServiceBus.RavenDB/blob/70fd44638395148d9acbcb01a1e2138f930a555c/src/NServiceBus.RavenDB/SagaPersister/SagaPersister.cs#L27
- https://github.com/Particular/NServiceBus.RavenDB/blob/70fd44638395148d9acbcb01a1e2138f930a555c/src/NServiceBus.RavenDB/SagaPersister/SagaPersister.cs#L66
- https://github.com/Particular/NServiceBus.RavenDB/blob/70fd44638395148d9acbcb01a1e2138f930a555c/src/NServiceBus.RavenDB/SagaPersister/SagaPersister.cs#L88

So I wonder if the `SageMetaModel` should expose such lookups ? Otherwise `SageMetaModel` is useless for Raven as well!

andreasohlund commented:
>Yes, I don't think is worth the hassle, but up to you

I think it's worth it (to move types users should care about into the roor `NServiceBus` namespace).

@Particular/nservicebus your thoughts?

danielmarbach commented:
I don’t agree about the ‘root namespace’ I think we should get our
structure of namespaces cleaned up. But I have no idea how to do this in a
nonbreaking way

SimonCropp commented:
@danielmarbach 

>  I think we should get our structure of namespaces cleaned up

what does this mean?

danielmarbach commented:
I'll let a picture speak

![nservicebus_structure](https://cloud.githubusercontent.com/assets/174258/6892874/3f6fe8f6-d6c8-11e4-9198-c1c58a39a83b.png)

![nservicebus_tangle_1024](https://cloud.githubusercontent.com/assets/174258/6892875/3f846e20-d6c8-11e4-8053-8450193a6669.png)

SimonCropp commented:
@danielmarbach those picture maen nothing to me in the context of namespaces. namespaces mean nothing in terms of structure or complexity.

johnsimons commented:
@andreasohlund I think this is done, please review and do any modifications that you think need done.

johnsimons commented:
Actually I think `CorrelationProperty`, `SagaMetadata`, `SagaFinderDefinition` and `SagaMessage` should be immutable, thoughts ?

andreasohlund commented:
+1 for making them immutable

johnsimons commented:
@andreasohlund can you please review the description and title of this issue.
Also what issues does this PR closes ?

andreasohlund commented on src/NServiceBus.Core/Saga/UniqueAttribute_obsolete.cs:
```diff
@@ -17,19 +18,10 @@ public sealed class UniqueAttribute : Attribute
         /// </summary>
         /// <param name="type">The <see cref="Type"/> to evaluate.</param>
         /// <returns>A <see cref="PropertyInfo"/> of the property marked with a <see cref="UniqueAttribute"/> or null if not used.</returns>
+        [ObsoleteEx(RemoveInVersion = "7.0", TreatAsErrorFromVersion = "6.0", Message = "Use the new SagaMetadata")]
```
Since the entire class is obsoleted do we need obsoletes on each individual method/prop/field? (just curios cc @SimonCropp )

SimonCropp replied:
should be enough to obsolete the class. unless the methods/props have more explicit alternatives and u want to show a different compile error

andreasohlund replied:
Thanks for clarifying @simoncropp

On Mon, Apr 13, 2015 at 1:47 PM, Simon Cropp <notifications@github.com>
wrote:

> In src/NServiceBus.Core/Saga/UniqueAttribute_obsolete.cs
> <https://github.com/Particular/NServiceBus/pull/2693#discussion_r28231556>
> :
>
> > @@ -17,19 +18,10 @@ public sealed class UniqueAttribute : Attribute
> >          /// </summary>
> >          /// <param name="type">The <see cref="Type"/> to evaluate.</param>
> >          /// <returns>A <see cref="PropertyInfo"/> of the property marked with a <see cref="UniqueAttribute"/> or null if not used.</returns>
> > +        [ObsoleteEx(RemoveInVersion = "7.0", TreatAsErrorFromVersion = "6.0", Message = "Use the new SagaMetadata")]
>
> should be enough to obsolete the class. unless the methods/props have more
> explicit alternatives and u want to show a different compile error
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/2693/files#r28231556>.
>
